agent/dns.go

```diff
@@ -1358,13 +1361,11 @@ func (d *DNSServer) preparedQueryLookup(cfg *dnsConfig, network, datacenter, que
 		d.serviceNodeRecords(cfg, out.Datacenter, out.Nodes, req, resp, ttl, maxRecursionLevel)
 	}
 
-	d.trimDNSResponse(cfg, network, req, resp)
-
 	// If the answer is empty and the response isn't truncated, return not found
-	if len(resp.Answer) == 0 && !resp.Truncated {
-		d.addSOA(cfg, resp)
-		return
+	if len(resp.Answer) == 0 {
+		return errNoAnswer
 	}
+	return nil
```

Comment:
Looking at this part of the code changes, I'm wondering if the handling of SOA is still correct in this branch. There is a similar branch above in `serviceLookup`.

Previously we would only set SOA if we had no answers, where as now we'll set it in both cases (when the prepared query or service lookup returns no instances, and when it returns some service instances).

@blake do you know in what cases would we not want to set the `SOA` record on the response.NS section? Only for recurse, or are there other cases where setting the SOA record would be incorrect?

Reply:
I changed the code a bit to avoid trimming the answer after adding SOA, we only add SOA in certain conditions + no trimming happened (otherwise we will bust the response size limit).

Reply:
Ok, I looked into this again and I think it was mostly correct. The only problem I found was that my fix in the commit with title "dns: correct rcode for qtype not supported" I returned `errNoAnswer` (which sets SOA) instead of `nil` (which doesn't). They both use the same RCode, but only one will set SOA.

I have no idea if this is correct, but at least it matches the old behaviour. I've amended the commit which introduced the problem to properly return a `nil` error when the Rcode is not supported by `nodeLookup`.

Reply:
For unsupported qtypes, I believe the better response is to send `errNoAnswer` (or `errNoData` as I suggested renaming to).

https://umbrella.cisco.com/blog/nxdomain-nodata-debugging-dns-dual-stacked-hosts offers a good explanation as to why a NODATA response should be returned if resource records exist for other query types. Section 2.2 from RFC 2308 (linked in another comment) explains why its recommended to add the SOA.

> @blake do you know in what cases would we not want to set the SOA record on the response.NS section? Only for recurse, or are there other cases where setting the SOA record would be incorrect?

It took me a bit of staring at the code before I realized that response.Ns actually refers  to the the authority section of the DNS response. [[1]] :-)

After reading through a few RFCs, the SOA should definitely be omitted in negative responses for non-authoritative resources (i.e., recursive queries). It is also recommended that the SOA be added to the authority section [[2]] of a negative response (NODATA or NXDOMAIN) for resources which the the server is authoritative [[3]].

> I have no idea if this is correct,…

Based on my understanding of the linked RFCs, this seems to be the correct behavior.

[1]: https://pkg.go.dev/github.com/miekg/dns#Msg
[2]: https://datatracker.ietf.org/doc/html/rfc2181#section-7.1
[3]: https://datatracker.ietf.org/doc/html/rfc1034#section-4.3.4